typescript/lib/networks/network_context.ts

```diff
@@ -0,0 +1,440 @@
+
+            // destFp, err := os.Create(destAbsFilepath)
+            // if err != nil {
+            //     return stacktrace.Propagate(err, "An error occurred opening static file '%v' destination file '%v' for writing", staticFileId, destAbsFilepath)
+            // }
+            // defer destFp.Close()
+
+            // if _, err := io.Copy(destFp, srcFp); err != nil {
+            //     return stacktrace.Propagate(err, "An error occurred copying all the bytes from static file '%v' source filepath '%v' to destination filepath '%v'", staticFileId, srcAbsFilepath, destAbsFilepath)
+            // }
+
+        }
+        return null;
+    }
+
+    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
+    public registerFilesArtifacts(filesArtifactUrls: Map<FilesArtifactID, string>): Error {
+        const filesArtifactIdStrsToUrls: Map<string, string> = new Map();
+        for (let artifactId in filesArtifactUrls) {
+            const url: string = filesArtifactUrls[artifactId];
+            filesArtifactIdStrsToUrls[String(artifactId)] = url;
+        }
+        const args: RegisterFilesArtifactsArgs = newRegisterFilesArtifactsArgs(filesArtifactIdStrsToUrls);
+        // TODO TODO TODO - CALLBACK & ERROR-HANDLING
+        // if _, err := networkCtx.client.RegisterFilesArtifacts(context.Background(), args); err != nil {
+        //     return stacktrace.Propagate(err, "An error occurred registering files artifacts: %+v", filesArtifactUrls)
+        // }
+        return null;
+    }
+
+    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
+    public addService(
+        serviceId: ServiceID,
+        containerCreationConfig: ContainerCreationConfig, //TODO
+        generateRunConfigFunc: (ipAddr: string, generatedFileFilepaths: Map<string, string>, staticFileFilepaths: Map<StaticFileID, string>) => [ContainerRunConfig, Error]
+    ): [ServiceContext, Map<string, PortBinding>, Error] { //TODO
+
+        const [serviceContext, hostPortBindings, err] = this.AddServiceToPartition(
+            serviceId,
+            defaultPartitionId,
+            containerCreationConfig,
+            generateRunConfigFunc,
+            )
+        if (err != null) {
+            return [null, null, err ] //TODO no personalized error message
+        }
+
+        return [serviceContext, hostPortBindings, null]
+    }
+
+    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
+    public AddServiceToPartition(
+        serviceId: ServiceID,
+        partitionId: PartitionID,
+        containerCreationConfig: ContainerCreationConfig,
+        generateRunConfigFunc: (ipAddr: string, generatedFileFilepaths: Map<string, string>, staticFileFilepaths: Map<StaticFileID, string>) => [ContainerRunConfig, Error],
+    ): [ServiceContext, Map<string, PortBinding>, Error] {
+
+        //ctx := context.Background()
+        const log: logger.Logger = new logger.Logger();
```

Comment:
Good instinct in finding `tslog`! We'll probably want to use Winston ( https://github.com/winstonjs/winston ) though, as this seems to be what the rest of the world uses (17k stars vs tslog's 500).

So _generally_, the way logging frameworks work is that you'll have a global method like `winston.info` or `winston.debug`, just like we use `logrus.info` or `logrus.debug`.

When a framework does that, the logger is usually configured at a global level in the `main` function of the application, and then all the things flow from that.

However, it didn't look like `tslog` does that - you just create a logger object and pass it around everywhere. That's totally fine too (and IMO is less magic than a global log variable/configuration). In that alternate reality where we stayed with `tslog`, what we'd want to do is make the Logger a property of `NetworkContext`, and pass it in at construction time. In this way, you'd create a `Logger` object in the `main` function, pass it down to all the classes you create, and it would sort of "flow" through your program.

Of course, that has the downside of needing to pass that `Logger` object basically _everywhere_, which is why logging frameworks like Winston and Logrus generally have the global methods.

For Typescript, I recommend we a) switch to Winston and b) stay with the global pattern